src/PowerShellEditorServices/Session/Host/TerminalPSHostRawUserInterface.cs

```diff
@@ -184,11 +145,87 @@ public override Size MaxWindowSize
         /// <returns>A KeyInfo struct with details about the current keypress.</returns>
         public override KeyInfo ReadKey(ReadKeyOptions options)
         {
-            Logger.Write(
-                LogLevel.Warning,
-                "PSHostRawUserInterface.ReadKey was called");
+            // Converts ConsoleKeyInfo objects to KeyInfo objects and caches key down events
+            // for the next key up request.
+            KeyInfo ProcessKey(ConsoleKeyInfo key, bool isDown)
+            {
+                // Translate ConsoleModifiers to ControlKeyStates
+                ControlKeyStates states = default;
+                if ((key.Modifiers & ConsoleModifiers.Alt) != 0)
+                {
+                    states |= ControlKeyStates.LeftAltPressed;
+                }
+
+                if ((key.Modifiers & ConsoleModifiers.Control) != 0)
+                {
+                    states |= ControlKeyStates.LeftCtrlPressed;
+                }
+
+                if ((key.Modifiers & ConsoleModifiers.Shift) != 0)
+                {
+                    states |= ControlKeyStates.ShiftPressed;
+                }
 
-            throw new System.NotImplementedException();
+                var result = new KeyInfo((int)key.Key, key.KeyChar, states, isDown);
+                if (isDown)
+                {
+                    this.lastKeyDown = result;
+                }
+
+                return result;
+            }
+
+            bool includeUp = (options & ReadKeyOptions.IncludeKeyUp) != 0;
+
+            // Key Up was requested and we have a cached key down we can return.
+            if (includeUp && this.lastKeyDown != null)
+            {
+                KeyInfo info = this.lastKeyDown.Value;
+                this.lastKeyDown = null;
+                return new KeyInfo(
+                    info.VirtualKeyCode,
+                    info.Character,
+                    info.ControlKeyState,
+                    keyDown: false);
+            }
```

Comment:
KeyUp is requested, but why are you returning a KeyDown, especially that this KeyDown has already been returned previously?

Reply:
re: [emulating `ConsoleHost` behavior](https://github.com/PowerShell/PowerShellEditorServices/pull/874#discussion_r264380810)